ecosystem/python/sdk/aptos_sdk/account_sequence_number.py

```diff
@@ -0,0 +1,178 @@
+# Copyright © Aptos Foundation
+# SPDX-License-Identifier: Apache-2.0
+import asyncio
+import logging
+import time
+from typing import Optional
+
+from aptos_sdk.account_address import AccountAddress
+from aptos_sdk.async_client import RestClient
+
+
+class AccountSequenceNumber:
+    """
+    A managed wrapper around sequence numbers that implements the trivial flow control used by the
+    Aptos faucet:
+    * Submit up to 50 transactions per account in parallel with a timeout of 20 seconds
+    * If local assumes 50 are in flight, determine the actual committed state from the network
+    * If there are less than 50 due to some being committed, adjust the window
+    * If 50 are in flight Wait .1 seconds before re-evaluating
+    * If ever waiting more than 30 seconds restart the sequence number to the current on-chain state
+    Assumptions:
+    * Accounts are expected to be managed by a single AccountSequenceNumber and not used otherwise.
+    * They are initialized to the current on-chain state, so if there are already transactions in
+      flight, they make take some time to reset.
+    * Accounts are automatically initialized if not explicitly
+
+    Notes:
+    * This is co-routine safe, that is many async tasks can be reading from this concurrently.
+    * The synchronize method will create a barrier that prevents additional next_sequence_number
+      calls until it is complete.
+    * This only manages the distribution of sequence numbers it does not help handle transaction
+      failures.
+    """
+
+    client: RestClient
+    account: AccountAddress
+    lock = asyncio.Lock
+
+    maximum_in_flight: int = 100
+    maximum_wait_time = 30
+    sleep_time = 0.01
+
+    last_committed_number: Optional[int]
+    current_number: Optional[int]
+
+    def __init__(self, client: RestClient, account: AccountAddress):
+        self.client = client
+        self.account = account
+        self.lock = asyncio.Lock()
+
+        self.last_uncommitted_number = None
+        self.current_number = None
+
+    async def next_sequence_number(self, block: bool = True) -> Optional[int]:
+        """
+        Returns the next sequence number available on this account. This leverages a lock to
+        guarantee first-in, first-out ordering of requests.
+        """
+        await self.lock.acquire()
```

Comment:
Nit: I'd recommend using this as a context manager to make it impossible to forget to release.